src/core/shelf.sol

```diff
@@ -80,54 +80,104 @@ contract Shelf is TitleOwned {
-        shelf[loan].initial = principal_;
+    // --- Shelf: Loan actions ---
+    function issue(address registry, uint token) public returns (uint) {
+        require(NFTLike(registry).ownerOf(token) == msg.sender, "nft-not-owned");
+
+        bytes32 nft = keccak256(abi.encodePacked(registry, token));
+        require(nftlookup[nft] == 0, "nft-in-use");
+
+        uint loan = title.issue(msg.sender);
+        nftlookup[nft] = loan;
+
+        shelf[loan].registry = registry;
+        shelf[loan].tokenId = token;
+
+        return loan;
     }
 
-    // Move the NFT out of the shelf
-    function move(uint loan, address registry_, uint nft_, address to) public owner(loan) {
-        NFTLike(registry_).transferFrom(address(this), to, nft_);
+    function close(uint loan) public owner(loan) {
+        require(pile.debt(loan) == 0, "outstanding-debt"); // TODO: only allow closing of a loan that isn't active anymore. maybe there is a better criteria
+        title.close(loan);
+        bytes32 nft = keccak256(abi.encodePacked(shelf[loan].registry, shelf[loan].tokenId));
+        nftlookup[nft] = 0;
     }
 
-    function release (uint loan, address usr) public owner(loan) {
-        uint debt = pile.debtOf(loan);
-        require(debt == 0, "debt");
-        move(loan, shelf[loan].registry, shelf[loan].tokenId, usr);
-        adjust(loan);
+    // --- Currency actions ---
+    function want() public view returns (int) {
+        return int(balance) - int(tkn.balanceOf(address(this))); // safemath
     }
 
-    function free(uint loan, address usr) public auth  {
-        NFTLike(shelf[loan].registry).transferFrom(address(this), usr, shelf[loan].tokenId);
-        adjust(loan);
+    function borrow(uint loan, uint wad) public owner(loan) {
+        require(nftLocked(loan), "nft-not-locked");
+        ceiling.borrow(loan, wad);
+        pile.accrue(loan);
+        pile.incDebt(loan, wad);
+
+        balances[loan] = add(balances[loan], wad);
+        balance += wad;
     }
 
-    function deposit (uint loan, address usr) public owner(loan) {
+    function withdraw(uint loan, uint wad, address usr) public owner(loan) note {
+        require(nftLocked(loan), "nft-not-locked");
+        require(wad <= balances[loan], "amount-too-high");
+        balances[loan] -= wad;
+        balance -= wad;
+        tkn.transferFrom(address(this), usr, wad);
+    }
+
+    function repay(uint loan, uint wad) public owner(loan) note {
+        require(nftLocked(loan), "nft-not-locked");
+        require(balances[loan] == 0,"before repay loan needs to be withdrawn");
+        doRepay(loan, msg.sender, wad);
+    }
+
+    function recover(uint loan, address usr, uint wad) public auth {
+        doRepay(loan, usr, wad);
+        // TODO: do we need to decrease the loan balance here? There might be some
+        uint loss = pile.debt(loan);
+        pile.decDebt(loan, loss);
+    }
+
+    function doRepay(uint loan, address usr, uint wad) internal {
```

Comment:
I would propose we adopt the openzeppelin convention here: prefixing `_` to internal methods:
```suggestion
    function _repay(uint loan, address usr, uint wad) internal {
```

Reply:
Just created a refactoring/styling ticket to collect all the changes we want to implement for v2. https://github.com/centrifuge/tinlake/issues/172
@lucasvo @charleenfei @manuelpolzhofer let's add all the ideas/request there

Reply:
It's much harder to find these kind of things later on. I really would recommend to just fix these things right away.

Reply:
k, integrated into this pr